rfcs/0008-storage-ref.md

```diff
@@ -0,0 +1,249 @@
+
+```rust
+pub trait AsStorageRef<T> {
+    fn as_storage_ref(self) -> StorageRef<T>;
+}
+```
+
+Calling `as_storage_ref` is only valid for a storage accesses. A storage access looks like `storage.a.b.c...`. For example:
+
+```rust
+struct MyStruct {
+    x: u64,
+    y: b256,
+}
+
+storage {
+    s: MyStruct, 
+}
+
+fn foo() {
+    let x_ref: StorageRef<u64> = storage.s.x.as_storage_ref();
+    let y_ref: StorageRef<b56> = storage.s.y.as_storage_ref();
+    let s_ref: StorageRef<MyStruct> = storage.s.as_storage_ref();
+}
+```
+
+In order to use a storage reference, one can call `store` and `get` as follows:
+```rust
+let x_ref: StorageRef<u64> = storage.s.x.as_storage_ref();
+x_ref.store(42); // equivalent to `storage.x = 42`
+let x = x_ref.get(); // equivalent to `let x = storage.s.x`
+
+let y_ref: StorageRef<u64> = storage.s.y.as_storage_ref();
+y_ref.store(ZERO_B256); // equivalent to `storage.y = ZERO_B256`
+let y = y_ref.get(); // equivalent to `let y = storage.s.y`
+
+let s_ref: StorageRef<MyStruct> = storage.s.as_storage_ref();
+s_ref.store(MyStruct { x: 42, y: ZERO_B256 } ); // equivalent to `storage.s = MyStruct { x: 42, y: ZERO_B256 }`
+let s = s_ref.get(); // equivalent to `let s = storage.s`
+```
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+One way of implementing `as_storage_ref` is via the `__get_storage_key` intrinsic:
+
+```rust
+impl AsStorageRef<T> for T {
+    #[inline(always)]
+    fn as_storage_ref() -> StorageRef<T> {
+        StorageRef {
+            key: __get_storage_key()
+        }
+    }
+}
+```
+
+Currently, the intrinsic just returns the storage key corresponding to the top level storage variable. It should be extended to return the storage key of an arbitrary storage accesses `storage.a.b.c...`.
+
+## Implementing `StorageMap` using `StorageRef`.
+
+The current implementation of `insert` and `get` for `StorageMap` rely on the intrinsic `__get_storage_key` directly. Instead, the implementation should use `StorageRef` directly, potentially as follows:
+
+```rust
+/// A persistent key-value pair mapping struct.
+pub struct StorageMap<K, V> {}
+
+impl<K, V> StorageMap<K, V> {
+    #[storage(write)]
+    pub fn insert(self: StorageRef<Self>, key: K, value: V) {
+        let key = sha256((key, self.key));
+        store::<V>(key, value);
+    }
+
+    #[storage(read)]
+    pub fn get(self: StorageRef<Self>, key: K) -> Option<V> {
+        let key = sha256((key, self.key));
+        get::<V>(key)
+    }
+
+    #[storage(write)]
+    pub fn remove(self: StorageRef<Self>, key: K) -> bool {
+        let key = sha256((key, self.key));
+        clear::<V>(key)
+    }
+}
+```
+
+This should still allow writing the following:
+
+```rust
+storage {
+    my_map: StorageMap<u64, u64> = StorageMap {}
+}
+
+my_map.insert(0, 0);
+let x = my_map.get(0);
+my_map.as_storage_ref().remove(0);
```

Comment:
Not sure I follow this line - is the `as_storage_ref()` necessary/intended here?

Reply:
So I'd like methods of `StorageMap` to work with and without the `as_storage_ref`, which is why I'm suggesting https://github.com/FuelLabs/sway/issues/4047.